janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMultiQueryStrategy.java

```diff
@@ -0,0 +1,121 @@
+import org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphTraversalUtil;
+import org.janusgraph.graphdb.tinkerpop.optimize.step.JanusGraphMultiQueryStep;
+import org.janusgraph.graphdb.tinkerpop.optimize.step.MultiQueriable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * @author Marko A. Rodriguez (https://markorodriguez.com)
+ * @author Matthias Broecheler (http://matthiasb.com)
+ */
+public class JanusGraphMultiQueryStrategy extends AbstractTraversalStrategy<TraversalStrategy.ProviderOptimizationStrategy> implements TraversalStrategy.ProviderOptimizationStrategy {
+
+    private static final Set<Class<? extends ProviderOptimizationStrategy>> PRIORS = new HashSet<>(Arrays.asList(JanusGraphLocalQueryOptimizerStrategy.class, JanusGraphStepStrategy.class));
+    private static final JanusGraphMultiQueryStrategy INSTANCE = new JanusGraphMultiQueryStrategy();
+
+    private JanusGraphMultiQueryStrategy() {
+    }
+
+    @Override
+    public void apply(final Admin<?, ?> traversal) {
+        if (!traversal.getGraph().isPresent()
+            || TraversalHelper.onGraphComputer(traversal)
+            // The LazyBarrierStrategy is not allowed to run on traversals which use drop(). As a precaution,
+            // this strategy should not run on those traversals either, because it can also insert barrier().
+            || !TraversalHelper.getStepsOfAssignableClassRecursively(DropStep.class, traversal).isEmpty()) {
+            return;
+        }
+
+        final StandardJanusGraph janusGraph = JanusGraphTraversalUtil.getJanusGraph(traversal);
+        if (janusGraph == null || !janusGraph.getConfiguration().useMultiQuery()) {
+            return;
+        }
+
+        insertMultiQuerySteps(traversal, janusGraph.getConfiguration().limitBatchSize());
+        configureMultiQueriables(traversal);
+    }
+
+    /**
+     * Insert JanusGraphMultiQuerySteps everywhere in the current traversal where MultiQueriable steps could benefit
+     *
+     * @param traversal The local traversal layer.
+     */
+    private void insertMultiQuerySteps(final Admin<?, ?> traversal, boolean limitBatchSize) {
+        JanusGraphTraversalUtil.getSteps(JanusGraphTraversalUtil::isMultiQueryCompatibleStep, traversal).forEach(step -> {
+            Optional<Step> multiQueryPosition = JanusGraphTraversalUtil.getLocalMultiQueryPositionForStep(step);
+            if (multiQueryPosition.isPresent() && JanusGraphTraversalUtil.isLegalMultiQueryPosition(multiQueryPosition.get())) {
+                Step pos = multiQueryPosition.get();
+                if (limitBatchSize && !(multiQueryPosition.get() instanceof NoOpBarrierStep)) {
+                    NoOpBarrierStep barrier = new NoOpBarrierStep(traversal);
+                    TraversalHelper.insertBeforeStep(barrier, pos, traversal);
+                    pos = barrier;
```

Comment:
Do you know how come L78 - L80 are not covered by tests? Seems they are executed as long as limitBatchSize is enabled, and no explicit NoOpBarrierStep is present.

Reply:
This was not covered indeed as all tests which had `query.limit-batch-size=true` also had explicitly set barriers. I added four extra test cases in `JanusGraphMultiQueryStrategyTest`.